### PR TITLE
OWCollocations: widget for observing collocations

### PR DESCRIPTION
**Pull request by ajdapretnar**
##### Issue

No widget for showing collocations.

##### Description of changes
OWCollocations widget.

Screenshot 2022-05-06 at 14 43 36

##### Includes
- [X] Code changes
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by ajdapretnar** [comment]
This is now ready for a review and comments.
TODO:
- [ ] It'd be nice if the widget showed colored bars, like Rank, but not sure how to do it. Thought I'd automatically get it by importing TableView from Rank.
- [ ] Documentation.
- [ ] Report.

**Comment by noahnovsak** [comment]
Adding colored bars could be done by adding `view.setItemDelegateForColumn(1, gui.ColoredBarItemDelegate())` in the widget's `__init__`, as `TableRank` overrides it for columns 0 and 1. Also `self.collModel.setExtremesFrom(1, scores)` would need to be called at some point after computing the scores.

If I understand what this widget does correctly, I imagine you could end up with quite a few columns here for larger datasets so running the computation concurrently makes sense. Also with radio buttons you can only view one set of scores at a time.
If these issues are not a concern then ignore this, otherwise it seems to me that some more functionality could easily be copied over from OWRank.

**Comment by ajdapretnar** [comment]
> I imagine you could end up with quite a few rows here for larger datasets

Most likely. I'd leave concurrent implementation to masters. :)

> Also with radio buttons you can only view one set of scores at a time.

This is intentional. The top words can vary significantly. I thought it makes more sense to view a single score at a time. Otherwise you'd have extremely long tables with many nans.

File: orangecontrib/text/widgets/owcollocations.py
Comment (PrimozGodec) [inline orangecontrib/text/widgets/owcollocations.py:162]:
```suggestion
```

I think it is unnecessary since you are creating a new table with only Collocations and Scores columns, so you know those two names are unique. Input table columns are not used in the new table.

File: orangecontrib/text/widgets/owcollocations.py
Comment (PrimozGodec) [inline orangecontrib/text/widgets/owcollocations.py:5]: It is not nice to import classes from widgets since they can be changed dramatically without previous deprecations (we do not assume that classes inside the widget file are used anywhere else). 

Since there is not so much code also used in this widget, I suggest copying TableModela and TableView from the rank widget and removing unused code.

File: orangecontrib/text/widgets/owcollocations.py
Comment (PrimozGodec) [inline orangecontrib/text/widgets/owcollocations.py:176]:
```suggestion
        self.collView.horizontalHeader().setSortIndicator(-1, Qt.AscendingOrder)
        if model is not None:
            model.resetSorting(yes_reset=True)
```

Resetting the order does not remove the indicator in the header. You can remove it by adding the line suggested above.

**Review by PrimozGodec — CHANGES_REQUESTED** [review]
It looks perfect. I just added a few minor comments.

File: orangecontrib/text/widgets/owcollocations.py
Comment (ajdapretnar) [reply orangecontrib/text/widgets/owcollocations.py:5]: I agree. I think both classes can be used elsewhere (what does Extract Keywords use?). I can prepare the PR and clean this one.

Comment (ajdapretnar) [reply orangecontrib/text/widgets/owcollocations.py:5]: @PrimozGodec Is this what you had in mind? https://github.com/biolab/orange3/pull/6098

Comment (PrimozGodec) [reply orangecontrib/text/widgets/owcollocations.py:5]: I had in mind to take the implementation from the rank and copy it into the Collocations widget. However, your proposed solution is excellent if other widgets use the same implementation (maybe Score Documents and Keywords can).

The only problem is that we would need to wait for Orange's release before we can merge this PR.

Comment (ajdapretnar) [reply orangecontrib/text/widgets/owcollocations.py:5]: Right. I can do either. I think having a nice table capable of showing certain scores as colored bars might be useful. @janezd @markotoplak @VesnaT What do you think?

**Comment by ajdapretnar** [comment]
@PrimozGodec As discussed, I've copied the code from Rank for now. Will submit another PR with changes from the master, which can be merged after we release Text and once Orange is released, too.

Other suggestions are also implemented.

**Comment by janezd** [comment]
If you really mean to someday remove the code from rank, you need to add a time bomb, that is, a test that will fail when Orange version is above XX, with a message to remove the code and this test.

**Comment by ajdapretnar** [comment]
You are absolutely right. I've added the test, hope it's ok.

**Review by PrimozGodec — APPROVED** [review]

**Comment by PrimozGodec** [comment]
Nicely done. :)

**Pull request merged**
